File: ios/Classes/DeviceInfo.swift
import Foundation
import Flutter
class DeviceInfo : NSObject{
    static var prefix = "DeviceInfo."
    
    static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let range = call.method.range(of: DeviceInfo.prefix){
            let method = call.method[range.upperBound...]
            if method == "model"{
                do{
                    let model = try platform();
                    result(model)
                }catch{
                    result(UIDevice.current.model)
                }
            }else if method == "manufacturer"{
                result("Apple")
            }
            else if method == "name"{
                result(UIDevice.current.name)
            }
            else if method == "version"{
                result(UIDevice.current.systemVersion)
            }
            else if method == "deviceType"{
                #if targetEnvironment(simulator)
                result("Virtual")
                #else
                result("Physical")
                #endif
            }
            else if method == "devicePlatform"{
                result("iOS")
            }
            else if method == "idiom"{
                switch(UIDevice.current.userInterfaceIdiom){
                case .phone:
                    result("Phone")
                case .pad:
                    result("Tablet")
                case .tv:
                    result("TV")
                case .carPlay, .unspecified:
                    result("Unknown")
                }
            }
            else{
                result(FlutterMethodNotImplemented)
            }
        }else{
            result(FlutterMethodNotImplemented)
        }
        
    }
    
    static func platform() throws -> String {
        var size = 0
        sysctlbyname("hw.machine", nil, &size, nil, 0)
        var machine = [CChar](repeating: 0,  count: size)
        sysctlbyname("hw.machine", &machine, &size, nil, 0)
        return String(cString: machine)
    }
}

File: lib/device_info.dart
import 'flutter_essentials.dart';

class DeviceInfo {
  static const String _methodPrefix = "DeviceInfo.";

  static Future<String> get model async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}model');
    return result;
  }

  static Future<String> get manufacturer async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}manufacturer');
    return result;
  }

  static Future<String> get name async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}name');
    return result;
  }

  static Future<String> get version async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}version');
    return result;
  }

  static Future<DeviceType> get deviceType async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}deviceType');
    return _stringToDeviceType(result);
  }

  static Future<DevicePlatform> get devicePlatform async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}devicePlatform');
    return _stringToDevicePlatform(result);
  }

  static Future<DeviceIdiom> get idiom async {
    final String result =
        await SHARED_CHANNEL.invokeMethod('${_methodPrefix}idiom');
    return _stringToDeviceIdiom(result);
  }

  static DevicePlatform _stringToDevicePlatform(String platform) {
    if (platform == "Android") {
      return DevicePlatform.Android;
    }

    if (platform == "iOS") {
      return DevicePlatform.iOS;
    }

    if (platform == "Web") {
      return DevicePlatform.Web;
    }

    return DevicePlatform.Unknown;
  }

  static DeviceType _stringToDeviceType(String type) {
    if (type == "Physical") {
      return DeviceType.Physical;
    }

    if (type == "Virtual") {
      return DeviceType.Virtual;
    }

    return DeviceType.Unknown;
  }

  static DeviceIdiom _stringToDeviceIdiom(String idiom) {
    if (idiom == "Phone") {
      return DeviceIdiom.Phone;
    }
    if (idiom == "Tablet") {
      return DeviceIdiom.Tablet;
    }
    if (idiom == "Desktop") {
      return DeviceIdiom.Desktop;
    }

    return DeviceIdiom.Unknown;
  }
}

enum DeviceType { Unknown, Physical, Virtual }
enum DevicePlatform { Android, iOS, Web, Unknown }
enum DeviceIdiom { Phone, Tablet, Desktop, Unknown }
